C2 does not do constant-folding of xor

j3graham opened the pull request:
C2 does not eliminate XOR nodes when it's arguments are constants. This has a noticeable effect on `Long.expand` on architectures that don't have instructions equivalent `PDEP` to be used in an intrinsic.

This patch demonstrates a potential fix to the problem, but there might well be better ways to do it.

j3graham commented:
Sample program:
```
public class XorTest {
    public static void main(String[] args) {
        long t=0;
        for (int i = 0; i < 10_000; i++) {
            t+=doXor();
        }
        System.out.println("t = " + t);
    }

    static long doXor(){
        long c=42;
        return c ^ 2025L;
    }
}
```

IGV before patch (constant nodes are red)

![Xor before](https://github.com/user-attachments/assets/8cecb845-8273-4647-91fa-919dec02a4f0)


IGV after patch
![25 Return](https://github.com/user-attachments/assets/36f02474-df70-4a0c-b1c0-71c02732501f)

SirYwell commented:
The code for this optimization seems to be present already in the `add_ring` methods, and from my understanding those should be called by `AddNode::Value`. Does that mean `XorINode::Value` and `XorLNode::Value` never call that method in your tests, and instead always end up in the `if` directly below your changes? Maybe that `if` should rather check to ignore constant values?

liach commented:
Created https://bugs.openjdk.org/browse/JDK-8347645 for you. The causes are #2776 and #4136.

I think the right way forward is to remove XorI/LNode::Value and move the code to their add_ring, which has lower priority than constant folding, instead. However I am not a hotspot/JIT engineer, so don't take my words for granted.

eme64 commented:
@j3graham Since this is a performance improvement: do you have any benchmark that shows a speedup?

eme64 reviewed (commented):
Thanks for working on this!

I have not yet looked at the VM cpp changes yet.

Some comments about the test:
Please move it to:
`test/hotspot/jtreg/compiler/c2/gvn/TestXor.java`

The comments sometimes mention `c3` etc, but there may only be a `c` or `x`. Please fix them ;)

The tests should also do result verification. Currently you only check that we have the expected nodes, but constant folding could have bugs we would not catch this way. What I usually do:
Compute some `GOLDEN`, which should be computed in interpreter, and then with a `@Check` method you can compare the result to that `GOLDEN` value.

Plus: it would be nice if the constants could be picked at random. You can do that with a `public static final int CON = random_value`.

Best would be if you could use the new `Generators`, see
`./test/hotspot/jtreg/compiler/lib/generators/Generators.java`

Let me know if you need any more help with that.

j3graham commented:
Some rough performance numbers with existing tests:

Aarch64 M4 Pro
Baseline:
```
Benchmark       (size)  Mode  Cnt  Score   Error  Units
Longs.compress     500  avgt    5  0.512 ? 0.002  us/op
Longs.expand       500  avgt    5  0.503 ? 0.014  us/op
```

This PR:
```
Benchmark       (size)  Mode  Cnt  Score   Error  Units
Longs.compress     500  avgt    5  0.420 ? 0.008  us/op
Longs.expand       500  avgt    5  0.312 ? 0.002  us/op
```

It's sensitive to what the mask value is. With the mask from https://github.com/openjdk/jdk/pull/22928, the numbers are:
Baseline:
`Longs.expand2      500  avgt   15  0.494 ? 0.002  us/op`

This PR:
`Longs.expand2      500  avgt    5  0.242 ? 0.002  us/op
`

liach commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -1004,14 +991,29 @@ const Type *XorINode::add_ring( const Type *t0, const Type *t1 ) const {
+    // Complementing a boolean?
+    if( r0 == TypeInt::BOOL && ( r1 == TypeInt::ONE
+                                 || r1 == TypeInt::BOOL))
+      return TypeInt::BOOL;
```
@eme64 Can you verify that it's right for the bounded value handling to come before this boolean range checking, and the bounded value handling is still correct if one of the input arguments is constant?

eme64 replied:
@j3graham can you give us an explanation here?

liach replied:
This is how the existing code worked - the bounded handling came before this bool handling.  This bool restriction has been here since the earliest publication of OpenJDK, I think.  I believe a professional with gvn knows better about this than maybe j3graham, who seems not as familiar.

j3graham replied:
I have removed the boolean check, per note from @merykitty

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -1004,14 +991,29 @@ const Type *XorINode::add_ring( const Type *t0, const Type *t1 ) const {
-  if( r0 == TypeInt::BOOL && ( r1 == TypeInt::ONE
-                               || r1 == TypeInt::BOOL))
-    return TypeInt::BOOL;
+  if( !r0->is_con() || !r1->is_con() ) {
```
```suggestion
  if( !r0->is_con() || !r1->is_con()) {
```
That is the new style, please apply it everywhere in the code.

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -1004,14 +991,29 @@ const Type *XorINode::add_ring( const Type *t0, const Type *t1 ) const {
 
-  if( !r0->is_con() || !r1->is_con() ) // Not constants
-    return TypeInt::INT;        // Any integer, but still no symbols.
+    // result of xor can only have bits sets where any of the
+    // inputs have bits set. lo can always become 0.
+
+    if ((r0->_lo >= 0) &&
+        (r0->_hi > 0)  &&
+        (r1->_lo >= 0) &&
+        (r1->_hi > 0)) {
+      // hi - set all bits below the highest bit. Using round_down to avoid overflow.
+      const TypeInt* t1x = TypeInt::make(0, round_down_power_of_2(r0->_hi) + (round_down_power_of_2(r0->_hi) - 1), r0->_widen);
+      const TypeInt* t2x = TypeInt::make(0, round_down_power_of_2(r1->_hi) + (round_down_power_of_2(r1->_hi) - 1), r1->_widen);
+      return t1x->meet(t2x);
+    }
```
I know you only moved this code - but can you add some more comments about what it does? Maybe add some examples? It would help me review this code faster if I don't have to figure it out myself ;)

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -1022,11 +1024,26 @@ const Type *XorLNode::add_ring( const Type *t0, const Type *t1 ) const {
+      const TypeLong* t1x = TypeLong::make(0, round_down_power_of_2(r0->_hi) + (round_down_power_of_2(r0->_hi) - 1), r0->_widen);
+      const TypeLong* t2x = TypeLong::make(0, round_down_power_of_2(r1->_hi) + (round_down_power_of_2(r1->_hi) - 1), r1->_widen);
+      return t1x->meet(t2x);
+    }
```
Is there a reason why the `BOOL` check is not here as well? Does `long` not have one as well? This could be a follow-up RFE of course.

j3graham replied:
Long doesn't have a BOOL. Happily since the boolean test is covered by the more general check, we're good here.

eme64 reviewed (changes requested):
@j3graham Thanks for working on this.

I see you added some constant-folding tests, thanks for randomizing the constants and fixing the comments!

You do not just touch constant-folding code, but also the `BOOL` case and other non-constant cases. Do we have adequate testing for those? I'm especially asking because we are now changing the order of those optimizations - and it would be a shame if we got a regression because of it and did not catch if because we have no tests.

Maybe we do have tests, then you can just point me to them. Otherwise, it would be great if you could add some more IR tests for those cases as well :)

Thanks again!

merykitty commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -1004,14 +991,29 @@ const Type *XorINode::add_ring( const Type *t0, const Type *t1 ) const {
+    if ((r0->_lo >= 0) &&
+        (r0->_hi > 0)  &&
+        (r1->_lo >= 0) &&
+        (r1->_hi > 0)) {
```
This can be simplified to:

    if (r0->_lo >= 0 && r1->_lo >= 0)

There is no need to check `_hi` here.

j3graham replied:
If one arg was (lo=0,hi=1) and the other was 0 though, it might be a problem. Depends where x ^ 0 gets simplified.

j3graham replied:
*edit2* Actually, the alternative version has an issue if either of the lo values is 0

j3graham replied:
How about this?
    ``juint max = round_up_power_of_2<juint>( (r0->_hi | r1->_hi) +1) - 1;``

merykitty replied:
Why? `r0->_hi | r1->_hi == 0` would imply `r0->_hi == 0 && r1->_hi == 0`. This means that both operands are constants and should be filtered out from the previous step already.

j3graham replied:
The symptom I'm seeing with the alternative version is that the build fails with an exception of 
```
Creating jdk.net.jmod
Error: More than one SourceFile attribute
java.lang.module.InvalidModuleDescriptorException: More than one SourceFile attribute
	at java.base/jdk.internal.module.ModuleInfo.invalidModuleDescriptor(ModuleInfo.java:1228)
```
I am thinking that's a result of a bad xor. If I change the ``if ( (r0->_lo >= 0) `` to ``if ( (r0->_lo > 0)`` the build works again, but that might just be luck.

The case I'm concerned about is if `r0->_hi `and `r1->_hi` are both the same power of 2, say for example 4 and both lo values are 0. Then `round_up_power_of_2(4,4)-1` is 3. However, if the runtime value of one arg is 0 and the other is 4, `0 ^ 4 = 4, which is larger than 3.

merykitty commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -1004,14 +991,29 @@ const Type *XorINode::add_ring( const Type *t0, const Type *t1 ) const {
+      return t1x->meet(t2x);
+    }
+
+    // Complementing a boolean?
```
This is subsumed by the more-general check above and can be removed.

merykitty commented on test/hotspot/jtreg/compiler/c2/gvn/TestXorInt.java:
```diff
@@ -0,0 +1,83 @@
+/*
```
We have `compiler/c2/irTests/XorINodeIdealizationTest.java`, is adding a separate test file necessary?

j3graham commented:
Remaining issues are
- do we keep the original code that calculates the the bounds, or substitute something that looks easier to explain.
- do we omit the check for r0->_hi and r1->_h1 being > 0. Without it r0=(0,0) and r1=(0,1) , gets past the check that both are constant with r0->_hi equal to zero.  In the original code this would cause a problem - except that 0^x should have been simplified already, so it would be prevented. But that code is a little further away, so not immediately clear.

j3graham commented:
I have added tests specifically for the power of 2 case, and updated `add_ring` to use the simpler logic. I've also added some examples to help clarify it.

After https://github.com/openjdk/jdk/pull/17508, this might lend itself to an even more intuitive solution.

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -1004,16 +991,29 @@ const Type *XorINode::add_ring( const Type *t0, const Type *t1 ) const {
+    return TypeInt::make( r0->get_con() ^ r1->get_con() );
+  }
+
+  // not constants
```
Well I suppose one of them could still be a constant... You should make your comments here a little more precise ;)

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -1004,16 +991,29 @@ const Type *XorINode::add_ring( const Type *t0, const Type *t1 ) const {
+      // x cannot have any bit set that is higher than the highest bit set in r0->_hi
+      // y cannot have any bit set that is higher than the highest bit set in r1->_hi
+
+      // note cast to unsigned happens before +1 to avoid signed overflow, and
+      // round_up is safe because high bit is unset (0 <= lo <= hi)
```
```suggestion
      // Note: cast to unsigned happens before +1 to avoid signed overflow, and
      // round_up is safe because high bit is unset (0 <= lo <= hi)
```

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -1004,16 +991,29 @@ const Type *XorINode::add_ring( const Type *t0, const Type *t1 ) const {
+
+      // note cast to unsigned happens before +1 to avoid signed overflow, and
+      // round_up is safe because high bit is unset (0 <= lo <= hi)
+      juint max = round_up_power_of_2(juint(r0->_hi | r1->_hi) + 1) - 1;
```
Ok, changes like these will require a lot more testing. Now we have to deal with all sorts of ranges. Imagine we now have some tighter range `r = [a..b]`, but we get it off a little. Then a if-branch could wrongly constant fold badly:
`if (a < r)`
This would seem always `false`.

So we need some Java tests that validate the ranges we produce here.

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -1004,16 +991,29 @@ const Type *XorINode::add_ring( const Type *t0, const Type *t1 ) const {
-  // Otherwise just XOR them bits.
-  return TypeInt::make( r0->get_con() ^ r1->get_con() );
+  if (r0->_lo >= 0 && r1->_lo >= 0) {
+      // x ^ y cannot have any bit set that is higher than both the highest bits set in x and y
```
You should have a 2-indentation here, not 4-indentation ;)

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -1004,16 +991,29 @@ const Type *XorINode::add_ring( const Type *t0, const Type *t1 ) const {
+      return TypeInt::make(0, max, MAX2(r0->_widen, r1->_widen));
+  }
+
+  return TypeInt::INT;        // Any integer, but still no symbols.
```
I know you copied this. But what does it mean "no symbols"? Feel free to just remove the comment, it's not really helpful :sweat_smile: 
```suggestion
  return TypeInt::INT;
```

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -1022,12 +1022,46 @@ const Type *XorLNode::add_ring( const Type *t0, const Type *t1 ) const {
-  if( !r0->is_con() || !r1->is_con() )
-    return TypeLong::LONG;      // Any integer, but still no symbols.
+  if (r0->is_con() && r1->is_con()) {
+    // just XOR them bits.
```
```suggestion
    // Constant fold: (c1 ^ c2) -> c3
```

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -1022,12 +1022,46 @@ const Type *XorLNode::add_ring( const Type *t0, const Type *t1 ) const {
+      // x cannot have any bit set that is higher than the highest bit set in r0->_hi
+      // y cannot have any bit set that is higher than the highest bit set in r1->_hi
+
+      // we want to find a value that has all 1 bits everywhere up to and including
```
Can you please capitalize the first word in your sentences? ;)

eme64 commented on test/hotspot/jtreg/compiler/c2/irTests/XorLNodeIdealizationTests.java:
```diff
@@ -33,6 +33,11 @@
+
+    private static final long CONST_1 = RunInfo.getRandom().nextLong();
+    private static final long CONST_2 = RunInfo.getRandom().nextLong();
+    private static final long CONST_POW_2 = Math.abs(1L << RunInfo.getRandom().nextInt());
```
Can you please use our new `Generators.java`? I'll explain why:

Sampling random powers-of-2, or values near powers-of-2 is very very rare with longs. There are about 64 powers-of-2 in a set of `2^64`, so the odds are really really bad. The sampler in `Generators.java` has a bias to pick more powers-of-2 and values nearby.

eme64 replied:
And we need values nearby powers of 2, as your examples show ;)

eme64 replied:
I'm a little hesitant to modify old tests. So I'd suggest to just do it for the new ones :smile:

eme64 commented on test/hotspot/jtreg/compiler/c2/irTests/XorLNodeIdealizationTests.java:
```diff
@@ -217,4 +229,57 @@ public long test16(long x, long y, long z) {
+        return x ^ x;
+    }
+
+    // clamp value to [1,CONST_POW_2]
+    @ForceInline
+    private static long forceMinMax(long value) {
+        // equivalent to Math.min(CONST_POW_2, Math.max(value, 1))
+        return 1L + (value & (CONST_POW_2 - 1L));
+    }
+
+    @Test
+    @IR(counts = {IRNode.XOR, "1"})  // must not be constant-folded
+    // checks that add_ring computes correct max on exact powers of 2
+    public boolean testMaxPow2(long x, long y) {
+        x = forceMinMax(x);
+        y = forceMinMax(y);
+        long xor = x ^ y;
+        return xor < CONST_POW_2;
+    }
+
+    @DontCompile
+    public boolean interpretedMaxPow2(long x, long y) {
+        x = forceMinMax(x);
+        y = forceMinMax(y);
+
+        long xor = x ^ y;
+        return xor < CONST_POW_2;
+    }
+
+    @Test
+    @IR(failOn = {IRNode.XOR})
+    @IR(counts = {IRNode.CON_I, "1"})
+    public boolean testMaxPow2Folded(long x, long y) {
+        x = forceMinMax(x);
+        y = forceMinMax(y);
+
+        long xor = x ^ y;
+        return xor < (CONST_POW_2*2L);
+    }
```
Thanks for adding range-tests like these!
I think we need some smarter tests on top of these though.

For one, we should have some camped ranges that are not restricted to `[1,CONST_POW_2]`, but have any lower and upper limit.

Further: we should check with `if` branches if the resulting range is correct. Not that we get the resulting range off-by-1 and have some edge case that constant-folds the wrong way ;)

eme64 commented on test/hotspot/jtreg/compiler/c2/irTests/XorINodeIdealizationTests.java:
```diff
@@ -33,6 +33,12 @@
  * @run driver compiler.c2.irTests.XorINodeIdealizationTests
  */
 public class XorINodeIdealizationTests {
+    private static final int CONST_1 = RunInfo.getRandom().nextInt();
```
You'll also have to update the copyright of the tests here ;)

eme64 reviewed (commented):
@j3graham Thanks for the updates! It's looking better already.

I think the VM code is generally looking quite reasonalbe (modulo some formatting and comments).

We need to work in improving the tests a little, especially because you changed the VM code. I left some comments below :)

Good work, keep it up!

merykitty commented:
I would advise against adding more complicated jtreg tests as reasoning about them can be tricky. For these operations I think it would be more preferable to use a gtest which can directly inspect the details of `Type` objects while jtreg tests can only reason about them indirectly. You can generate some pairs of `Type`s, calculate their xor, then, try to sample some values inside the operands and assert that the results would be a member of the resulted `Type`.

eme64 commented:
@merykitty @j3graham 
> I would advise against adding more complicated jtreg tests as reasoning about them can be tricky. For these operations I think it would be more preferable to use a gtest which can directly inspect the details of Type objects while jtreg tests can only reason about them indirectly. You can generate some pairs of Types, calculate their xor, then, try to sample some values inside the operands and assert that the results would be a member of the resulted Type.

Yes, actually `gtest` may be a better solution if that works!

j3graham commented on test/hotspot/jtreg/compiler/c2/irTests/XorINodeIdealizationTests.java:
```diff
@@ -217,4 +233,74 @@ public int test16(int x, int y, int z) {
+    // clamp value to [1,CONST_POW_2]
+    @ForceInline
+    private static int forceMinMax(int value){
+        // equivalent to Math.min(CONST_POW_2, Math.max(value, 1))
+        return 1 + (value & (CONST_POW_2 - 1));
```
Surprisingly the min/max bounds do not get propagated on longs. Once I found a pattern that worked there I used it for ints as well, though min/max is ok there.

eme64 replied:
Ah, I think that is because we do not yet intrinsify `Long.max/min` yet! See https://github.com/openjdk/jdk/pull/20098.

j3graham commented:
I'm going to need some guidance on using gtest for this. I can't really call xor on the types - that would need a Node which doesn't seem to easily be created in a stand-alone way. I can move the tricky code from add_ring to a static method to test that more directly. I'm having trouble getting a test to run without the vm crashing.

As soon as I  try creating a type, for example `const TypeLong *v=TypeLong::make(0)`, I get something like this:

`

merykitty commented:
That's unfortunate :( You can proceed with the jtreg tests.

eme64 commented:
That is unfortunate indeed. It would be nice if we had some way of running the Types independently of the CompilerThread. It would really improve our ability to test Types.

j3graham commented:
By narrowing down the scope of my testable static method, I've now managed to write some gttests that exercise the most mysterious part of the code. I've also built up more in the IR tests - probably to the point where they have become too complicated, and have significant overlap with the gtests. So my feeling is that the IR tests could be thinned out, but some other opinions would be good.

merykitty commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -1004,30 +991,88 @@ const Type *XorINode::add_ring( const Type *t0, const Type *t1 ) const {
+  return TypeInt::INT;        // Any integer, but still no symbols.
 }
 
+jint XorINode::calc_xor_max(const jint hi_0, const jint hi_1) {
```
Since this is only used in `XorINode::Value`, I think a static local method would be better. And if you are comfortable with templates, I suggest templatizing it to unify the implementation with `jlong` below.

j3graham replied:
I've templatized it and took it out of the class, but didn't make it static, because I need the test to call it. Can I do better than that?

merykitty commented on test/hotspot/gtest/opto/test_xor_node.cpp:
```diff
@@ -0,0 +1,33 @@
+#include "unittest.hpp"
+#include "runtime/interfaceSupport.inline.hpp"
+
+TEST_VM(opto, xor) {
```
Did you forget to add the tests to this file?

j3graham replied:
Fixed.

merykitty commented on src/hotspot/share/opto/addnode.hpp:
```diff
@@ -253,6 +253,38 @@ class XorLNode : public AddNode {
 };
 
+template<class S, class U>
+static S calc_xor_max(const S hi_0, const S hi_1) {
```
If this is exposed in the header file, then I think it would be better to have it be a static function of a class to avoid convoluting the global namespace. Suggestion: Can you have `XorINode::calc_max` which will call a static `calc_xor_max` inside the cpp file.

j3graham replied:
I've reorganized it.

merykitty commented on src/hotspot/share/opto/addnode.hpp:
```diff
@@ -430,4 +462,7 @@ class MinDNode : public MaxNode {
   int min_opcode() const { return Op_MinD; }
 };
 
+template<class S, class U> S calc_xor_max(const S hi_0, const S hi_1);
```
This can probably removed.

merykitty commented on test/hotspot/gtest/opto/test_xor_node.cpp:
```diff
@@ -0,0 +1,94 @@
+}
+
+template <class S>
+void test_exhaustive_values(S hi_0, S hi_1){
```
You may want to test exhaustively for the `hi` values here, too. E.g:

    void test_exhaustive(S limit) {
        for (S hi0 = 0; hi0 <= limit; hi0++) {
            for (S hi1 = 0; hi1 <= limit; hi1++) {
                S max = calc_max(hi0, hi1);
                for (S v0 = 0; v0 <= hi0; v0++) {
                    for (S v1 = 0; v1 <= hi1; v1++) {
                        S v = v0 | v1;
                        EXPECT_LE(v, max);
                    }
                }
            }
        }
    }

j3graham replied:
Was my intention that the test should have been doing that. I've cleaned it up to clarify it.

merykitty commented on test/hotspot/gtest/opto/test_xor_node.cpp:
```diff
@@ -0,0 +1,94 @@
+void test_xor_bounds(S hi_0, S hi_1, S val_0, S val_1) {
+
+  // Skip out-of-bounds values for convenience
+  if(val_0> hi_0 || val_0 < S(0) || val_1 > hi_1 || val_1< S(0)) {
```
Please be consistent with the style.

merykitty commented on test/hotspot/jtreg/compiler/c2/irTests/XorINodeIdealizationTests.java:
```diff
@@ -33,16 +34,24 @@
+    private static final int CONST_1 = G.next();
+    private static final int CONST_2 = G.next();
+
+
```
Extra line break

merykitty commented:
Very nice, I think the patch looks good, please do another round of style refinement. In particular, make sure that there is no white space after `(` or before `)`, and after `if` or `for` we prefer having a whitespace before the `(`.

j3graham commented:
Thanks. I've done another round of format fixing. I've also simplified the IR tests so they don't try to cover as much as gtest does, and added equivalent tests for long.

I have temporarily left the more elaborate tests commented out in XorINodeIdealizationTests. I will remove them if nobody thinks they are worth keeping.

merykitty commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -1004,16 +1023,26 @@ const Type *XorINode::add_ring( const Type *t0, const Type *t1 ) const {
-    return TypeInt::BOOL;
+  if (r0->is_con() && r1->is_con()) {
+    // Constant fold: (c1 ^ c2) -> c3
+    return TypeInt::make( r0->get_con() ^ r1->get_con() );
```
Format

merykitty commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -1022,12 +1051,26 @@ const Type *XorLNode::add_ring( const Type *t0, const Type *t1 ) const {
-    return TypeLong::LONG;      // Any integer, but still no symbols.
+  if (r0->is_con() && r1->is_con()) {
+    // Constant fold: (c1 ^ c2) -> c3
+    return TypeLong::make( r0->get_con() ^ r1->get_con() );
```
Format here, too.

merykitty commented on test/hotspot/jtreg/compiler/c2/irTests/XorINodeIdealizationTests.java:
```diff
@@ -217,4 +226,164 @@ public int test16(int x, int y, int z) {
+        return xor < 0b1000;
+    }
+
+/*
```
Please remove these commented out code

merykitty reviewed (approved):
Otherwise LGTM, very nice tests, thanks very much!

eme64 commented:
@j3graham Can you please update the PR description at the top? The current version does not reflect the most up-to-date explanations, right?

I would like to see a nice summary, what cases were covered before and what cases you are now covering additionally.
Give a quick explanation how you changed the code.

I'll increase the number of reviewers as this looks like a substantial change.

/reviewers 2 reviewer

eme64 commented:
I also see that https://github.com/openjdk/jdk/pull/2776 and https://github.com/openjdk/jdk/pull/4136 were mentioned here. Both of those are related an have no IR tests of their own, yikes! We have to ensure that we cover those old cases, and then new ones here, so that we do not get any accidental regressions.

Maybe that's all already covered in other existing tests or the tests you added. Can you please provide a summary of all tests and what cases they cover in the PR description? It would help a lot for reviewing.

eme64 reviewed (changes requested):
@j3graham Thanks for taking this on! It's great to see someone clean this up and make sure all the cases optimize as expected!

merykitty commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -978,23 +978,42 @@ const Type* XorINode::Value(PhaseGVN* phase) const {
 
 
+template<class S, class U>
+static S calc_xor_max(const S hi_0, const S hi_1) {
```
I think the name is good enough, calculate the max of a xor is a pretty self-explanatory name. You just need a better description for the method. Suggestion:

    Given 2 non-negative values in the ranges [0, hi_0] and [0, hi_1], respectively. The bitwise xor of these values should also be non-negative. This method calculates its maximum.

eme64 replied:
What about calling it `calculate_upper_bound_of_xor_with_non_negative_inputs`?

j3graham replied:
I've gone with `calc_xor_upper_bound_of_non_neg`

merykitty commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -1004,16 +1023,26 @@ const Type *XorINode::add_ring( const Type *t0, const Type *t1 ) const {
+  return TypeInt::INT;
+}
+
+jint XorINode::calc_max(const jint hi_0, const jint hi_1)  {
+  return calc_xor_max<jint, juint>(hi_0, hi_1);
 }
```
`calc_xor_max` is a static method in the cpp file and thus it is not preferrable to import it directly into the test file.

eme64 replied:
I suppose this is more of a cosmetic concern, not as important. Up to you what you want to do @j3graham .

j3graham replied:
I tried a few variations, and this seemed the least invasive one. I miss java package-scope.

merykitty commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -1022,12 +1051,26 @@ const Type *XorLNode::add_ring( const Type *t0, const Type *t1 ) const {
-  return TypeLong::make( r0->get_con() ^ r1->get_con() );
+  // At least one of the arguments is not constant
+
+  // Result of xor can only have bits sets where any of the
+  // inputs have bits set. lo can always become 0.
+
```
I think the comment should be removed, the specs and implementation of `calc_xor_max` should be referred to when trying to understand this piece of code.

eme64 replied:
I think a better name like `calculate_upper_bound_of_xor_with_non_negative_inputs` would help a lot, and make some comments redundant.

j3graham replied:
I have removed the comment - it was left over from the original code.

j3graham commented:
I have updated the summary to be more informative. I believe the related optimizations are now covered in tests.

jaskarth commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -995,22 +995,42 @@ const Type* XorINode::Value(PhaseGVN* phase) const {
+  // Note: cast to unsigned happens before +1 to avoid signed overflow, and
+  // round_up is safe because high bit is unset (0 <= lo <= hi)
+
+  return round_up_power_of_2(U(hi_0 | hi_1) + 1) - 1 ;
```
```suggestion
  return round_up_power_of_2(U(hi_0 | hi_1) + 1) - 1;
```

jaskarth commented on test/hotspot/jtreg/compiler/c2/irTests/XorINodeIdealizationTests.java:
```diff
@@ -217,4 +226,101 @@ public int test16(int x, int y, int z) {
+    }
+
+    @Test
+    public int testXorConstRange(int x, int y) {
```
Should this have an `@IR` test attached, like `testFoldableXor`?

j3graham replied:
This test is really there to do a minimal correctness test of the xor calculation, rather than verifying the IR. The xor is not expected to be eliminated.

j3graham replied:
I've updated that test to make more sense, and include IR checks

jaskarth reviewed (commented):
This looks really nice! I just have 2 small comments here.

jaskarth reviewed (approved):
Thanks for the update! It looks good to me.

j3graham commented:
Hi, @eme64, so you gave any more comments on this PR?

eme64 commented on test/hotspot/jtreg/compiler/c2/irTests/XorINodeIdealizationTests.java:
```diff
@@ -217,4 +226,110 @@ public int test16(int x, int y, int z) {
+        Asserts.assertEQ(true, testFoldableXorPow2(a, b));
+    }
+
+    @Test
+    @IR(failOn = {IRNode.XOR})
+    @IR(counts = {IRNode.CON_I, "1"})
+    public boolean testFoldableXorPow2(int x, int y) {
+        return ((x & 0b1000) ^ (y & 0b1000)) < 0b10000;
+    }
+
+    @Test
+    @IR(counts = {IRNode.XOR, "1"})
+    public boolean testUnfoldableXorPow2(int x, int y) {
+        return ((x & 0b1000) ^ (y & 0b1000)) < 0b1000;
+    }
+
+    @Test
+    @IR(failOn = {IRNode.XOR})
+    @IR(counts = {IRNode.CON_I, "1"})
+    public boolean testFoldableXor(int x, int y) {
+        var xor = (x & 0b111) ^ (y & 0b100);
+        return xor < 0b1000;
+    }
```
FYI I'm working on fuzzing the compiler currently, so I'm thinking about how to write such tests more generally, that's why I took the time to come up with this. If you have better ideas, I'm more than happy to see them ;)

j3graham replied:
I started down the path of more elaborate randomization (see code removed in https://github.com/openjdk/jdk/pull/23089/commits/4a2912021103cefbe30fb3cc9e7d303b63ea454d). Instead of that approach. I went with doing the more detailed coverage with gtest, and having just a few specific checks in the IR tests. For example the "pow2" tests are there because that was a scenario that caused some trouble. I think there is a relatively small set of "interesting" cases that are nice to cover with something deterministic. (I could add a few more tests with hard-coded interesting values). It feels disproportionate to do something more complicated for this PR. 

That said, if there was more tooling in place to make more broad random tests easier to write, it would be very attractive. Looking at PR https://github.com/openjdk/jdk/pull/23418, it looks like that's what you've got in mind. Specifically, these kinds of tests with constants are begging for a way to spew out a whole bunch of related test methods.

eme64 replied:
> For example the "pow2" tests are there because that was a scenario that caused some trouble. I think there is a relatively small set of "interesting" cases that are nice to cover with something deterministic. (I could add a few more tests with hard-coded interesting values).

I know that there are only few "interesting" cases. That's why I came up with `Generators` so we would be more likely to hit the interesting cases.

Of course we want to have a good number of deterministic cases. But extending that with some randomized tests that would find bugs over a longer time span are still valuable.

The issue is often that there are edge cases, and humans are not very good at finding them. Thus with randomization we would at least eventually find those cases - hopefully. And in my experience I did find bugs with randomized tests. Sometimes we found bugs that were not related to the issue they were created for originally, but they found some other issue - and that is valuable too.

> It feels disproportionate to do something more complicated for this PR.

It's really not that complicated. All I'd be asking for is the test in the form that I provided above ;)

j3graham replied:
I've added the test.

eme64 reviewed (commented):
Thanks for the ping! Nice work, thanks for the improvements. I have a few more suggestions.

eme64 commented on test/hotspot/jtreg/compiler/c2/irTests/XorINodeIdealizationTests.java:
```diff
@@ -57,7 +59,7 @@ public void runMethod() {
         int d = RunInfo.getRandom().nextInt();
 
         int min = Integer.MIN_VALUE;
-        int max = Integer.MAX_VALUE;
+        int max = MAX_VALUE;
```
Let's make it consistent, and either use `Integer.` everywhere or without it everywhere ;)

j3graham commented:
Hi @eme64, do you have any more recommendations on this?

wenshao commented:
@merykitty Could you review the changed code?

eme64 commented on test/hotspot/jtreg/compiler/c2/irTests/XorINodeIdealizationTests.java:
```diff
@@ -332,4 +342,105 @@ public boolean testFoldableXor(int x, int y) {
+
+    @Test
+    public int testRandomLimits(int x, int y) {
+        x = RANGE_1.clamp(x);
+        y = RANGE_2.clamp(y);
```
Question:
did you verify that this `RANGE_1` with its clamp values are really detected as constants by C2, and not seen as loads? Maybe that works, but I've never tried it myself.

eme64 replied:
I  just hope that the abstraction here does not invalidate our intent to have constant min/max bounds for the clamping ;)

eme64 replied:
Ah, you are right, great :)

eme64 reviewed (changes requested):
Thanks for the updates @j3graham !

Could we have the same tests for long as for int? I see that the `testRandomLimits` only exists for int right now.

eme64 commented:
@j3graham something funny happened with the PR description, you may want to fix that:
![image](https://github.com/user-attachments/assets/b25b6a7c-24f6-4e90-a220-5ed94ce06f8b)

eme64 commented:
@j3graham I think the VM changes look good, and the tests are almost there. So I launched some testing. Please ping me in a day for the results!

j3graham commented:
Fixed the description and added randomized tests for Long. Happily min/max over Long is now optimized!

eme64 reviewed (approved):
Tests from yesterday passed. We only had minor changes, so I'll trust the GitHub Actions for testing.

@j3graham Thanks for your work on this, and all the updates during the review :blush:

j3graham commented:
Thank you @eme64, @merykitty, and @jaskarth for the feedback. I guess I’m now looking for one more Reviewer (with capital R)

liach commented:
@iwanowww Can you help review this, which causes expand and compress with constant mask not constant fold on aarch (x64 uses instructions)

iwanowww reviewed (commented):
Overall, looks good. Very nice unit test!

2 comments:
  * naming: why not simply `xor_upper_bound` instead of `calc_xor_upper_bound_of_non_neg`?
  * ceremony around exposing `calc_xor_upper_bound_of_non_neg` to gtest: have you tried to move it under `src/share/hotspot/utilities/` and include from both `addnode.cpp` and `test_xor_node.cpp`.

j3graham commented:
The naming of that method evolved during the course of the review of this PR. I believe the thinking was that the check was not necessarily an overall upper bound, and a simpler name would imply it was more general.

I have separated the logic out into its own header file, but I've put it under `opto`. `utilities` felt too far away from where it was used. I've called it `addnodeXorUtil.hpp` with the thinking that it could become the home for more code - it's a pretty small piece of code to have its own header.  Other naming suggestions would be welcomed.

iwanowww commented:
Thanks.

> The naming of that method evolved during the course of the review of this PR. I believe the thinking was that the check was not necessarily an overall upper bound, and a simpler name would imply it was more general.

There's usually a lot of invariants a function assumes and it's simply impractical to encode everything in the name. Speaking of this particular case (`calc_xor_upper_bound_of_non_neg`): 
  * `calc_` is redundant and IMO only adds noise;
  * `_non_neg` part is confusing; I'd stress instead that it works on **ranges**.

So, `xor_upper_bound_for_ranges` then? (And, please, explain in the comment what's the correspondense between `S` and `U` template type parameters.)

> `addnodeXorUtil.hpp`

I'm fine with placing it under `opto`. Please, rename the file into `src/hotspot/share/opto/utilities/xor.hpp`.

merykitty commented:
@iwanowww

> `_non_neg` part is confusing; I'd stress instead that it works on ranges.

I find it easier to think of it as calculating the upperbound of the xor of 2 non-negative integers whose upperbounds are given in the parameters.

j3graham commented:
Renamed to `xor_upper_bound_for_ranges` before I saw your comment, @merykitty. I'd be ok with another name though. With the last changes, the method is no longer a member of the class, so it's no longer going to get as many eyes on it without context, so maybe it matters less now.

iwanowww commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -1021,16 +1008,20 @@ const Type *XorINode::add_ring( const Type *t0, const Type *t1 ) const {
-  // Otherwise just XOR them bits.
-  return TypeInt::make( r0->get_con() ^ r1->get_con() );
+  if (r0->_lo >= 0 && r1->_lo >= 0) {
+      // Combine [0, lo_1] ^ [0, hi_1] -> [0, max]
```
What does this comment refer to? It mentions `lo_1` and `hi_1` while `r0->_hi` and `r1->_hi` are passed into `xor_upper_bound_for_ranges`. 

Also, I'd avoid naming it`max`: it sort of hints to `max_jint`, but in reality it represents the upper bound of the operation. Why not `upper_bound` instead?

iwanowww commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -1021,16 +1008,20 @@ const Type *XorINode::add_ring( const Type *t0, const Type *t1 ) const {
-                               || r1 == TypeInt::BOOL))
-    return TypeInt::BOOL;
+  if (r0->is_con() && r1->is_con()) {
+    // Constant fold: (c1 ^ c2) -> c3
```
A bit confusing. The comment mentions `c1` and `c2` while the code operate on `t0`/`r0` and `t1`/`r1`.

iwanowww reviewed (commented):
Overall, looks good.

A minor comments follow.

iwanowww reviewed (approved):
Looks good!

eme64 commented:
@j3graham I gave it a quick look, and it looks even better now. Let me run testing again before you integrate!

Please ping me in 24h for the results!

j3graham commented:
Hi @eme64, any news on test results?

eme64 reviewed (approved):
Testing looks good :)

j3graham commented:
/integrate

wenshao commented:
/sponsor

j3graham commented:
Thank you all for your help. I’m looking forward to https://github.com/openjdk/jdk/pull/17508 making these kinds of optimizations more systematic.
